eclair-core/src/main/scala/fr/acinq/eclair/payment/send/Recipient.scala

```diff
@@ -0,0 +1,178 @@
+ * limitations under the License.
+ */
+
+package fr.acinq.eclair.payment.send
+
+import fr.acinq.bitcoin.scalacompat.ByteVector32
+import fr.acinq.bitcoin.scalacompat.Crypto.PublicKey
+import fr.acinq.eclair.crypto.Sphinx
+import fr.acinq.eclair.payment.Invoice.ExtraEdge
+import fr.acinq.eclair.payment.OutgoingPaymentPacket._
+import fr.acinq.eclair.payment.{Bolt11Invoice, OutgoingPaymentPacket}
+import fr.acinq.eclair.router.Router.{ChannelHop, NodeHop, Route}
+import fr.acinq.eclair.wire.protocol.PaymentOnion.{FinalPayload, IntermediatePayload}
+import fr.acinq.eclair.wire.protocol.{GenericTlv, OnionRoutingPacket, PaymentOnionCodecs}
+import fr.acinq.eclair.{CltvExpiry, Features, InvoiceFeature, MilliSatoshi, MilliSatoshiLong, ShortChannelId}
+import scodec.bits.ByteVector
+
+/**
+ * Created by t-bast on 28/10/2022.
+ */
+
+sealed trait Recipient {
+  /** Id of the final receiving node. */
+  def nodeId: PublicKey
+
+  /** Total amount that will be received by the final receiving node. */
+  def totalAmount: MilliSatoshi
+
+  /** CLTV expiry that will be received by the final receiving node. */
+  def expiry: CltvExpiry
+
+  /** Features supported by the recipient. */
+  def features: Features[InvoiceFeature]
+
+  /** Edges that aren't part of the public graph and can be used to reach the recipient. */
+  def extraEdges: Seq[ExtraEdge]
+
+  /** Build a payment to the recipient using the route provided. */
+  def buildPayloads(paymentHash: ByteVector32, route: Route): Either[OutgoingPaymentError, PaymentPayloads]
+}
+
+object Recipient {
+  /** Iteratively build all the payloads for a payment relayed through channel hops. */
+  def buildPayloads(finalAmount: MilliSatoshi, finalExpiry: CltvExpiry, finalPayload: NodePayload, hops: Seq[ChannelHop]): PaymentPayloads = {
+    // We ignore the first hop since the route starts at our node.
+    hops.tail.reverse.foldLeft(PaymentPayloads(finalAmount, finalExpiry, Seq(finalPayload))) {
+      case (current, hop) =>
+        val payload = NodePayload(hop.nodeId, IntermediatePayload.ChannelRelay.Standard(hop.shortChannelId, current.amount, current.expiry))
+        PaymentPayloads(current.amount + hop.fee(current.amount), current.expiry + hop.cltvExpiryDelta, payload +: current.payloads)
+    }
+  }
+}
+
+/** A payment recipient that can directly be found in the routing graph. */
+case class ClearRecipient(nodeId: PublicKey,
+                          features: Features[InvoiceFeature],
+                          totalAmount: MilliSatoshi,
+                          expiry: CltvExpiry,
+                          paymentSecret: ByteVector32,
+                          extraEdges: Seq[ExtraEdge] = Nil,
+                          paymentMetadata_opt: Option[ByteVector] = None,
+                          nextTrampolineOnion_opt: Option[OnionRoutingPacket] = None,
+                          customTlvs: Seq[GenericTlv] = Nil) extends Recipient {
```

Comment:
I agree, but the issue is that you cannot know whether you missed some of them, since it's optional, so it can end up being more confusing than helpful...